torch/distributed/_shard/checkpoint/state_dict_loader.py

```diff
@@ -0,0 +1,234 @@
+                raise ValueError(
+                    f"Invalid checkpoint metadata for {fqn}, " +
+                    "expected TensorStorageMetadata but found {type(md)}"
+                )
+        else:
+            # This is actually hard to handle correctly
+            # If the value is not a tensor but any random obj,
+            # we cannot just write whatever memory it points to inplace
+            # the best we can to is to replace it with an object of the same type
+            bytes_io = io.BytesIO()
+            brr = BytesReadRequest(
+                bytes=bytes_io,
+                storage_key=fqn,
+            )
+            bytes_read_requests.append(brr)
+
+    return (bytes_read_requests, tensor_read_requests)
+
+
+def load_state_dict(
+    state_dict: Dict[str, Any],
+    storage_reader: StorageReader,
+) -> None:
+    """
+    Load a distributed state_dict in SPMD style.
+
+    Each rank will try to read the least amount of data necessary
+    to fullfill the requested `state_dict`.
+
+    When loading ShardedTensor instances, each rank only
+    reads data for their local shards.
+
+    All tensors in ``state_dict`` must be allocated on their
+    destination device prior to calling this function.
+
+    All non-tensor data is loaded using `torch.load()`.
+
+    Args:
+        state_dict (Dict[str, Any]) : The state_dict to load. Note that this
+            state dict will updated in places.
+        storage_reader (StorageReader): StorageReader used to load data from.
+
+    Returns:
+        None.
+
+    Examples
+        >>> my_model = MyModule()
+        >>> optimizer = Adagrad(my_model.parameters())
+        >>> model_state_dict = my_model.state_dict()
+        >>> fs_storage_loader = torch.distributed._shard.checkpoint.FileSystemLoader("/checkpoint/1")
+
+        >>> torch.distributed._shard.checkpoint.load_state_dict(
+        >>>     state_dict=model_state_dict,
+        >>>     storage_reader=fs_storage_loader,
+        >>> )
+
+        >>> # module.load_state_dict() functon might have customized steps
+        >>> # to flush the state_dict, must call them to
+        >>> # ensure the correct behavior
+        >>> my_model.load_state_dict(model_state_dict)
+    """
+
+    metadata = storage_reader.read_metadata()
+    bytes_read_requests, tensor_read_requests = _reshard_and_prepare_read_request(
+        state_dict=state_dict, metadata_from_storage=metadata
+    )
+    bytes_futures = storage_reader.read_bytes(bytes_read_requests)
+
+    tensor_futures = storage_reader.read_tensors(tensor_read_requests)
+
+    bytes_futures.wait()
+
+    # Addtional steps are required to convert the bytes to its original type
+    # Note that this is NOT inplace,
+    # it creating a new object and replace what's in the state dict
+    for req in bytes_read_requests:
+        fqn = req.storage_key
+        # Ensure the BytesIO is rewound
+        req.bytes.seek(0)
+        state_dict[fqn] = torch.load(req.bytes)
```

Comment:
What about defining 2 layers of APIs here: a lower-level API that requires `BytesIO`, and then a higher-level API on top of it that can accept anything and performs conversion to `BytesIO` on an as-needed basis by calling `torch.save()`?

For users that want to ensure backwards compatibility of checkpoints, and avoid possibly allowing arbitrary code execution vulnerabilities, this allows them to use the lower-level API and ensure that they cannot end up accidentally using `pickle`.

Reply:
> The alternative design would require significantly more work than just switching from torch.save to save_state_dict which would be detrimental to our users a stifle adoption.

This arguably is not quite true. The general idea is users should be able to choose how they want serde to be done via `state_dict()` and `load_state_dict()` - they are defined in models. Today unfortunately a state_dict is `Dict[str, Any]` in pytorch but there's nothing preventing the checkpoint layer to say if you want to use it, please convert non tensor objects to `BytesIO` or otherwise I'll reject you.

For general APIs like this, IMHO it's always better to start restricted and then loosened over time. PyTorch generally tends to do the opposite and favor "user adoption" first which incurs hacks in lower level of the whole stack which arguably would lead to unresolvable tech debt over time.

Reply:
The current API does what you suggest, which is to have the storage layer only deal with BytesIO and Tensor.

What we lack is pluggable serialization such that we can avoid pickle or have a more efficient variant of `torch.save` for tensors.

Designing this pluggable piece is not trivial and will take some time. I'm tracking this issue in https://github.com/pytorch/pytorch/issues/76441 as a follow up change after this one.